### PR TITLE
ENH: Add groupby().ngroup() method to count groups (#11642)

### LINKED ISSUE #11642: ENH: enumerate groups
Sometimes it's handy to have access to a distinct integer for each group.  For example, using the (internal) grouper:

```
>>> df = pd.DataFrame({"a": list("xyyzxy"), "b": list("ab"*3), "c": range(6)})
>>> df["group_id"] = df.groupby(["a","b"]).grouper.group_info[0]
>>> df
   a  b  c  group_id
0  x  a  0         0
1  y  b  1         2
2  y  a  2         1
3  z  b  3         3
4  x  a  4         0
5  y  b  5         2
```

This can be achieved in a number of ways but none of them are particularly elegant, esp. if we're grouping on multiple keys and/or Series.  Accordingly, after a brief discussion on gitter, I propose a new method `transform("enumerate")` which returns a Series of integers from 0 to ngroups-1 matching the order the groups will be iterated in.  In other words, we'll simply be applying the following map:

```
>>> m = {k: i for i, (k,g) in enumerate(df.groupby(["a","b"]))}
>>> m
{('x', 'a'): 0, ('y', 'b'): 2, ('y', 'a'): 1, ('z', 'b'): 3}
```

(Note this is only to shows the desired behaviour, and wouldn't be how it'd be implemented!)

### PR DESCRIPTION
**Pull request by dsm054**
This basically adds a method to give access to the coding returned by `grouper.group_info[0]`, i.e. the number of the group that each row is in.  This is a natural parallel to cumcount(), and while it's not the world's most important feature it's come in handy for me from time to time and deserves a public method, IMHO.
- [X] closes #11642
- [X] 9 tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -1331,6 +1331,74 @@ def nth(self, n, dropna=None):
 
     @Substitution(name='groupby')
     @Appender(_doc_template)
+    def enumerate(self, ascending=True):
+        """
+        Number each group from 0 to the number of groups - 1.
+
+        This is the enumerative complement of cumcount.  Note that the
+        numbers given to the groups match the order in which the groups
```
Comment (jreback) [inline pandas/core/groupby.py]: versionadded tag

File: pandas/tests/test_groupby.py [hunk 2]
```diff
@@ -5141,6 +5141,118 @@ def test_cumcount_groupby_not_col(self):
         assert_series_equal(expected, g.cumcount())
         assert_series_equal(expected, sg.cumcount())
 
+    def test_enumerate(self):
```
Comment (jreback) [inline pandas/tests/test_groupby.py]: can u out these in a separate class TestCounting or something (and maybe move cumcount there as well)

File: doc/source/groupby.rst [hunk 3]
```diff
@@ -980,6 +980,23 @@ To see the order in which each row appears within its group, use the
...
+method (compare with ``cumcount``):
+
+.. ipython:: python
+
+   df = pd.DataFrame(list('aaabba'), columns=['A'])
+   df
+
+   df.groupby('A').enumerate()
+
+   df.groupby('A').enumerate(ascending=False)  # kwarg only
```
Comment (jreback) [inline doc/source/groupby.rst]: can u add an example (maybe in the examples section of groupby.rst showing where this is useful)

**Comment by chris-b1** [comment]
I think you add this function [here](https://github.com/dsm054/pandas/blob/a6e60a7adfe0c2313d8678c77654cbdb55d00ba1/pandas/core/groupby.py#L98), so that `df.groupby(...).transform('enumerate')` also works.    Probably the same with `cumcount`.

File: doc/source/groupby.rst [hunk 4]
```diff
@@ -969,7 +969,7 @@ Enumerate group items
 .. versionadded:: 0.13.0
 
 To see the order in which each row appears within its group, use the
-``cumcount`` method:
+``cumcount`` method (compare with ``enumerate``):
```
Comment (jreback) [inline doc/source/groupby.rst]: not sure what this means

File: pandas/tests/test_groupby.py [hunk 5]
```diff
@@ -5141,6 +5141,118 @@ def test_cumcount_groupby_not_col(self):
...
+            g = df.groupby(['a'], sort=sort_flag)
+            df['group_id'] = -1
+            df['group_index'] = -1
+
+            for i, (key, group) in enumerate(g):
+                df.loc[group.index, 'group_id'] = i
+                for j, ind in enumerate(group.index):
+                    df.loc[ind, 'group_index'] = j
+
+            assert_series_equal(pd.Series(df['group_id'].values),
```
Comment (jreback) [inline pandas/tests/test_groupby.py]: test with ``transform`` as well

**Comment by sinhrks** [comment]
Isn't the name ``enumerate`` confusing? I expected something related to the iteration.

**Comment by jreback** [comment]
can you rebase / update?

**Comment by jreback** [comment]
looking back on this, shouldn't we call this ``g.factorize()``?

**Comment by dsm054** [comment]
@jreback: I'll buy that.  The collision with `enumerate` was always a bit unfortunate.

**Comment by jorisvandenbossche** [comment]
I agree `enumerate` can be confusing given that it is not in an iteration context here, but I find `factorize` also not ideal (give its notion of categoricals).

The verb is also 'number'. But that is maybe also a bit strange to use given the use of `number` as a noun? According to google translate/oxford dictionary:

> number: Mark with a number or assign a number to, typically to indicate position in a series:
>> Synonyms: assign a number to, categorize by number, specify by number, mark with a number, itemize, enumerate

So this points back to enumerate ..

**Comment by jreback** [comment]
how about ``.categorize()``? or ``.group_number()``?

File: pandas/core/groupby.py [hunk 6]
```diff
@@ -1331,6 +1331,74 @@ def nth(self, n, dropna=None):
...
+        dtype: int64
+        >>> df.groupby('A').enumerate(ascending=False)
+        0    1
+        1    1
+        2    1
+        3    0
+        4    0
+        5    1
+        dtype: int64
+        >>> df = pd.DataFrame([['b'], ['a'], ['a'], ['b']], columns=['A'])
```
Comment (MaximilianR) [inline pandas/core/groupby.py]: Or without the `list` expansion is fine too

**Comment by jreback** [comment]
any more thoughts on the name? does SQL call this anything?

summary (and some new)
- ``.factorize()``
- ``.enumerate()``
- ``.categorize()``
- ``.group_number()``
- ``.ngroup()``
- ``.number()``?

**Comment by jreback** [comment]
any further thoughts on this?

**Comment by jreback** [comment]
@dsm054 thoughts on this?

**Comment by dsm054** [comment]
Looking back I still think that `enumerate` was a poor choice of name because we're not returning an iterable of tuples, and factorize and categorize suggest contracts that I don't think we necessarily want to commit to (namely, that the numbers and/or types will match those that pandas would give otherwise).

`group_number` is explicit, at least, and `ngroup` has a certain parallel with `ngroups` which is kind of appealing.  `number` feels too open-ended to me.  

In SQL I'd use something like DENSE_RANK() to get this.

**Comment by jreback** [comment]
I like ``.ngroup()``

File: doc/source/groupby.rst [hunk 7]
```diff
@@ -1087,6 +1087,23 @@ To see the order in which each row appears within its group, use the
...
+method:
+
+.. ipython:: python
+
+   df = pd.DataFrame(list('aaabba'), columns=['A'])
+   df
+
+   df.groupby('A').ngroup()
+
+   df.groupby('A').ngroup(ascending=False)  # kwarg only
```
Comment (jorisvandenbossche) [inline doc/source/groupby.rst]: You can remove the "kwarg only" I think. It's not really clear to what it points, and it is also not correct I think

File: doc/source/groupby.rst [hunk 8]
```diff
@@ -1178,3 +1195,20 @@ column index name will be used as the name of the inserted column:
...
    result.stack()
+
+Multi-column factorization
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By using ``.ngroup()``, we can extract information about the groups in a
+way similar to ``pd.factorize()``, but which applies naturally to multiple
+columns of mixed type and different sources:
+
+.. ipython::python
```
Comment (jorisvandenbossche) [inline doc/source/groupby.rst]: space after the `::` needed

File: pandas/tests/groupby/test_groupby.py [hunk 9]
```diff
@@ -4304,6 +4251,192 @@ def test_cummin_cummax(self):
...
+
+        g_ngroup = g.ngroup()
+        g_cumcount = g.cumcount()
+        expected_ngroup = pd.Series([0, 1, 2, 0, 3])
+        expected_cumcount = pd.Series([0, 0, 0, 1, 0])
+
+        assert_series_equal(g_ngroup, expected_ngroup)
+        assert_series_equal(g_cumcount, expected_cumcount)
+
+    def test_ngroup_cumcount_pair(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/groupby/test_groupby.py]: Are you testing here something specific? Or just to have more test cases?

**Comment by jorisvandenbossche** [comment]
I agree that `factorize` and `categorize` are not ideal (for me due to the link to categoricals, and the result is here is not a categorical). 
I first didn't like `ngroup`, as it looked *too much* as the existing `ngroups`, and it's not the same. But maybe that is not bad after all. `group_number` would be the most explicit. But OK with `ngroup` I think.

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good! (nice suite of tests!)

**Comment by jorisvandenbossche** [comment]
Just to be sure we have considered this: we have the question whether this should behave similar to `cumcount` (transformer) or rather similar to `size` (reducer)? 
But I suppose it is mainly the transforming case that is useful? 

In the second case we could always have both behaviours as `.ngroup()` vs `.transform('ngroup')`

File: doc/source/groupby.rst [hunk 10]
```diff
@@ -1087,6 +1087,23 @@ To see the order in which each row appears within its group, use the
...
+~~~~~~~~~~~~~~~~
+
+.. versionadded:: 0.20.0
+
+To see the ordering of the groups themselves, you can use the ``ngroup``
+method:
+
+.. ipython:: python
+
+   df = pd.DataFrame(list('aaabba'), columns=['A'])
```
Comment (jreback) [inline doc/source/groupby.rst]: name this df something else as it *might* be clobbering the other parts of the docs (.e.g dfg or something)

File: doc/source/groupby.rst [hunk 11]
```diff
@@ -1087,6 +1087,23 @@ To see the order in which each row appears within its group, use the
 
    df.groupby('A').cumcount(ascending=False)  # kwarg only
 
+Enumerate groups
+~~~~~~~~~~~~~~~~
+
+.. versionadded:: 0.20.0
+
+To see the ordering of the groups themselves, you can use the ``ngroup``
+method:
```
Comment (jreback) [inline doc/source/groupby.rst]: I would add a comment differentiating this from ``.cumcount`` (when to use which)

File: doc/source/groupby.rst (hunk 10 above)
Comment (jreback) [inline doc/source/groupby.rst]: alternatively, maybe better is to use the same example as cumcount (change if needed to make it easier to talk about), and then you can contrast it

File: doc/source/groupby.rst [hunk 12]
```diff
@@ -1178,3 +1195,20 @@ column index name will be used as the name of the inserted column:
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By using ``.ngroup()``, we can extract information about the groups in a
+way similar to ``pd.factorize()``, but which applies naturally to multiple
+columns of mixed type and different sources:
+
+.. ipython::python
+
+    df = pd.DataFrame({"A": [1, 1, 2, 3, 2], "B": list("aaaba")})
+
```
Comment (jreback) [inline doc/source/groupby.rst]: can you elaborate on *why* this is useful

File: doc/source/whatsnew/v0.20.0.txt [hunk 13]
```diff
@@ -75,6 +75,21 @@ Strings passed to ``DataFrame.groupby()`` as the ``by`` parameter may now refere
 
    df.groupby(['second', 'A']).sum()
 
+.. _whatsnew_0200.enhancements.groupby_ngroup:
+
+Groupby Group Numbers
+^^^^^^^^^^^^^^^^^^^^^
+
+A new groupby method ``ngroup``, parallel to the existing ``cumcount``, has been added to return the group order (:issue:`11642`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: ``.ngroup()`` and ``.cumcount()``

File: doc/source/whatsnew/v0.20.0.txt [hunk 14]
```diff
@@ -75,6 +75,21 @@ Strings passed to ``DataFrame.groupby()`` as the ``by`` parameter may now refere
...
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: add a ref back to the docs

File: pandas/core/groupby.py [hunk 15]
```diff
@@ -1365,6 +1365,62 @@ def nth(self, n, dropna=None):
...
     @Substitution(name='groupby')
     @Appender(_doc_template)
+    def ngroup(self, ascending=True):
+        """
+        Number each group from 0 to the number of groups - 1.
+
+        This is the enumerative complement of cumcount.  Note that the
+        numbers given to the groups match the order in which the groups
+        would be seen when iterating over the groupby object, not the
+        order they are first observed.
```
Comment (jreback) [inline pandas/core/groupby.py]: this comment is good, add something like this to the docs (in groupy.rst) where you show an example

File: pandas/core/groupby.py [hunk 16]
```diff
@@ -1365,6 +1365,62 @@ def nth(self, n, dropna=None):
...
+        dtype: int64
+        >>> df.groupby('A').ngroup(ascending=False)
+        0    1
+        1    1
+        2    1
+        3    0
+        4    0
+        5    1
+        dtype: int64
+        """
```
Comment (jreback) [inline pandas/core/groupby.py]: can you add an example with multple groupers

File: pandas/tests/groupby/test_groupby.py [hunk 17]
```diff
@@ -4304,6 +4251,192 @@ def test_cummin_cummax(self):
             tm.assert_series_equal(expected, result)
 
 
+class TestCounting(tm.TestCase):
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: perfect you moved them! even better, can you move to a separate file

``pandas/tests/groupby/test_counting.py`` (cumcount & ngroup). If you find other methods that are very relevant ok (note I wouldn't move .count()).

**Review by jreback — APPROVED** [review]
lgtm. some doc comments. ping when ready.

File: doc/source/whatsnew/v0.20.0.txt (hunk 13 above)
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.20.0.txt]: or even better: link to their docstring page, using ``` :meth:`~pandas.core.groupby.GroupBy.ngroup`  ``` (will only display the ngroup)

File: doc/source/groupby.rst (hunk 7 above)
Comment (dsm054) [inline doc/source/groupby.rst]: Ah, this was a copy/paste from the cumcount doc which I didn't remove.  I don't think it holds for that either, so I'll remove them both.

File: pandas/tests/groupby/test_groupby.py (hunk 9 above)
Comment (dsm054) [inline pandas/tests/groupby/test_groupby.py]: If you mean `test_ngroup_matches_cumcount`, yeah, it's just a specific two-column case to make sure they align.  If you mean `test_ngroup_cumcount_pair`, that's to make sure that the (ngroup, cumcount) pair you can assign to each row matches expectation.

File: doc/source/groupby.rst (hunk 10 above)
Comment (dsm054) [inline doc/source/groupby.rst]: It's already the same as the cumcount example, I think.

File: doc/source/whatsnew/v0.20.0.txt (hunk 13 above)
Comment (dsm054) [inline doc/source/whatsnew/v0.20.0.txt]: I think the grammar actually works better without the parentheses -- `type(df.groupby("a").ngroup)` is what gives "method", after all.  Should I do this to every mention of `ngroup` in this doc?

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: @dsm054 usually do this in the first / most prominent mention (so here is obviously good). you certainly can do it more. I would also add the ref for cumcount here.

File: doc/source/groupby.rst (hunk 10 above)
Comment (jreback) [inline doc/source/groupby.rst]: oh ok, just reading as a user want to know: what is the difference between these and when should I use one over other (which you explain in the doc-string, just add somewhere here)

File: pandas/tests/groupby/test_groupby.py (hunk 9 above)
Comment (jorisvandenbossche) [inline pandas/tests/groupby/test_groupby.py]: I meant the `test_ngroup_cumcount_pair`. It is just not really clear to me what the test specifically tries to test in addition to the other tests

**Comment by jreback** [comment]
can you rebase / update

**Comment by jreback** [comment]
@dsm054 if you can rebase. IIRC this was basically ready.

File: doc/source/groupby.rst [hunk 18]
```diff
@@ -1176,6 +1198,26 @@ Regroup columns of a DataFrame according to their sum, and sum the aggregated on
    df
    df.groupby(df.sum(), axis=1).sum()
 
+Multi-column factorization
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By using ``.ngroup()``, we can extract information about the groups in a
+way similar to ``pd.factorize()``, but which applies naturally to multiple
```
Comment (jreback) [inline doc/source/groupby.rst]: can you make this ``:func:`factorize` `` and point a ref link to the factorize section of docs (in reshaping IIRC0

File: doc/source/whatsnew/v0.20.2.txt [hunk 19]
```diff
@@ -21,6 +21,7 @@ Enhancements
 
 - Unblocked access to additional compression types supported in pytables: 'blosc:blosclz, 'blosc:lz4', 'blosc:lz4hc', 'blosc:snappy', 'blosc:zlib', 'blosc:zstd' (:issue:`14478`)
 - ``Series`` provides a ``to_latex`` method (:issue:`16180`)
+- A new groupby method :meth:`~pandas.core.groupby.GroupBy.ngroup`, parallel to the existing :meth:`~pandas.core.groupby.GroupBy.cumcount`, has been added to return the group order (:issue:`11642`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.2.txt]: add a ref to the docs you just created here

**Review by jreback — APPROVED** [review]
minor doc comments. lgtm otherwise. ping on green.

**Comment by jreback** [comment]
@dsm054 if you can update docs as indicated and rebase today would be great.

**Comment by dsm054** [comment]
@jreback: had to rebase this morning 'cause of that linting error you fixed.  I'll keep doing so until we're green.

**Comment by dsm054** [comment]
test_write_fspath_all[to_hdf-writer_kwargs3-tables] failed, which doesn't seem related.

**Comment by TomAugspurger** [comment]
> test_write_fspath_all[to_hdf-writer_kwargs3-tables] failed, which doesn't seem related.

yeah, I didn't realize that test was flaky, but I am able to reproduce it occasionally.

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dsm054 nice PR!

**Pull request merged**